cli/azd/cmd/completion.go

```diff
@@ -0,0 +1,188 @@
+			Short:                 "Generate bash completion script.",
+			DisableFlagsInUseLine: true,
+		},
+		ActionResolver: newCompletionBashAction,
+		OutputFormats:  []output.Format{output.NoneFormat},
+		DefaultFormat:  output.NoneFormat,
+		HelpOptions: actions.ActionHelpOptions{
+			Description: getCmdCompletionBashHelpDescription,
+			Footer:      getCmdCompletionBashHelpFooter,
+		},
+	})
+
+	completionGroup.Add("zsh", &actions.ActionDescriptorOptions{
+		Command: &cobra.Command{
+			Short:                 "Generate zsh completion script.",
+			DisableFlagsInUseLine: true,
+		},
+		ActionResolver: newCompletionZshAction,
+		OutputFormats:  []output.Format{output.NoneFormat},
+		DefaultFormat:  output.NoneFormat,
+		HelpOptions: actions.ActionHelpOptions{
+			Description: getCmdCompletionZshHelpDescription,
+			Footer:      getCmdCompletionZshHelpFooter,
+		},
+	})
+
+	completionGroup.Add("fish", &actions.ActionDescriptorOptions{
+		Command: &cobra.Command{
+			Short:                 "Generate fish completion script.",
+			DisableFlagsInUseLine: true,
+		},
+		ActionResolver: newCompletionFishAction,
+		OutputFormats:  []output.Format{output.NoneFormat},
+		DefaultFormat:  output.NoneFormat,
+		HelpOptions: actions.ActionHelpOptions{
+			Description: getCmdCompletionFishHelpDescription,
+			Footer:      getCmdCompletionFishHelpFooter,
+		},
+	})
+
+	completionGroup.Add("powershell", &actions.ActionDescriptorOptions{
+		Command: &cobra.Command{
+			Short:                 "Generate PowerShell completion script.",
+			DisableFlagsInUseLine: true,
+		},
+		ActionResolver: newCompletionPowerShellAction,
+		OutputFormats:  []output.Format{output.NoneFormat},
+		DefaultFormat:  output.NoneFormat,
+		HelpOptions: actions.ActionHelpOptions{
+			Description: getCmdCompletionPowerShellHelpDescription,
+			Footer:      getCmdCompletionPowerShellHelpFooter,
+		},
+	})
+}
+
+type completionAction struct {
+	shell string
+	cmd   *cobra.Command
+}
+
+func newCompletionBashAction(cmd *cobra.Command) actions.Action {
+	return &completionAction{shell: "bash", cmd: cmd}
+}
+
+func newCompletionZshAction(cmd *cobra.Command) actions.Action {
+	return &completionAction{shell: "zsh", cmd: cmd}
+}
+
+func newCompletionFishAction(cmd *cobra.Command) actions.Action {
+	return &completionAction{shell: "fish", cmd: cmd}
+}
+
+func newCompletionPowerShellAction(cmd *cobra.Command) actions.Action {
+	return &completionAction{shell: "powershell", cmd: cmd}
+}
+
+func (a *completionAction) Run(ctx context.Context) (*actions.ActionResult, error) {
+	rootCmd := a.cmd
+	for rootCmd.Parent() != nil {
+		rootCmd = rootCmd.Parent()
+	}
+
+	var err error
+	switch a.shell {
+	case "bash":
+		err = rootCmd.GenBashCompletion(a.cmd.OutOrStdout())
+	case "zsh":
+		err = rootCmd.GenZshCompletion(a.cmd.OutOrStdout())
+	case "fish":
+		err = rootCmd.GenFishCompletion(a.cmd.OutOrStdout(), true)
+	case "powershell":
+		err = rootCmd.GenPowerShellCompletion(a.cmd.OutOrStdout())
```

Comment:
nit. Make constants from the strings and use them when you add the subcommand and here